File: frontend/src/types/project.ts
import type { User } from '@/types/auth';

export type ProjectRole = 'owner' | 'member';

export interface Project {
  id: number;
  name: string;
  description: string | null;
  deadline: string | null; // ISO date "YYYY-MM-DD"
  owner_id: number;
  created_at: string;
  updated_at: string;
}

export interface ProjectMember {
  id: number;
  user: User;
  role: ProjectRole;
  joined_at: string;
}

export interface ProjectDetail extends Project {
  owner: User;
  members: ProjectMember[];
}

export interface ProjectCreatePayload {
  name: string;
  description?: string | null;
  deadline?: string | null;
}

export type ProjectUpdatePayload = Partial<ProjectCreatePayload>;

File: backend/app/schemas/project.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, EmailStr, Field

from app.models.enums import ProjectRole
from app.schemas.user import UserResponse


class ProjectBase(BaseModel):
    """Campos comunes de un proyecto."""

    name: str = Field(min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=2000)
    deadline: date | None = None


class ProjectCreate(ProjectBase):
    """Payload para crear un proyecto. El owner se toma del usuario autenticado."""


class ProjectUpdate(BaseModel):
    """Todos los campos opcionales: PATCH parcial."""

    name: str | None = Field(default=None, min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=2000)
    deadline: date | None = None


class ProjectResponse(ProjectBase):
    id: int
    owner_id: int
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ProjectMemberResponse(BaseModel):
    """Representación de un miembro dentro de un proyecto."""

    id: int
    user: UserResponse
    role: ProjectRole
    joined_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ProjectDetailResponse(ProjectResponse):
    """Proyecto con su lista de miembros (usado en GET /projects/{id})."""

    owner: UserResponse
    members: list[ProjectMemberResponse]


class ProjectMemberInvite(BaseModel):
    """Invitar a un miembro al proyecto por email.

    Buscar por email es más natural que pedir un user_id: el invitador
    no necesita conocer IDs internos.
    """

    email: EmailStr